kolibri/core/auth/management/commands/sync.py

```diff
@@ -172,103 +175,181 @@ def handle_async(self, *args, **options):
 
-        sync_client = network_connection.create_sync_session(
-            client_cert, server_cert, chunk_size=chunk_size
-        )
-
-        # pull from server and push our own data to server
-        if not no_pull:
-            self._setup_pull_progress_tracking(sync_client, noninteractive)
-            self._session_tracker_adapter(
-                sync_client.signals.session,
-                "Creating pull transfer session",
-                "Completed pull transfer session",
-            )
-            with db_task_write_lock:
-                sync_client.initiate_pull(Filter(dataset_id))
-        if not no_push:
-            self._setup_push_progress_tracking(sync_client, noninteractive)
-            self._session_tracker_adapter(
-                sync_client.signals.session,
-                "Creating push transfer session",
-                "Completed push transfer session",
-            )
-            with db_task_write_lock:
-                sync_client.initiate_push(Filter(dataset_id))
-
-        if not no_provision:
-            with db_task_write_lock:
-                create_superuser_and_provision_device(
-                    username, dataset_id, noninteractive=noninteractive
+        try:
+            # pull from server
+            if not no_pull:
+                self._handle_pull(
+                    network_connection,
+                    client_cert,
+                    server_cert,
+                    chunk_size,
+                    noninteractive,
+                    dataset_id,
+                )
+            # and push our own data to server
+            if not no_push:
+                self._handle_push(
+                    network_connection,
+                    client_cert,
+                    server_cert,
+                    chunk_size,
+                    noninteractive,
+                    dataset_id,
                 )
 
-        sync_client.close_sync_session()
+            if not no_provision:
+                with self._lock():
+                    create_superuser_and_provision_device(
+                        username, dataset_id, noninteractive=noninteractive
+                    )
+        except UserCancelledError:
+            if self.job:
+                self.job.extra_metadata.update(sync_state=State.CANCELLED)
+                self.job.save_meta()
+            logger.info("Syncing has been cancelled.")
+            return
+
+        network_connection.close()
 
         if self.job:
             self.job.extra_metadata.update(sync_state=State.COMPLETED)
+            self.job.save_meta()
 
         logger.info("Syncing has been completed.")
 
-    def _setup_pull_progress_tracking(self, sync_client, noninteractive):
+    @contextmanager
+    def _lock(self):
+        cancellable = False
+        # job can't be cancelled while locked
+        if self.job:
+            cancellable = self.job.cancellable
+            self.job.cancellable = False
```

Comment:
Only thing to flag here is that this will not change the value of cancellable transmitted to the frontend (as this will not get saved to the task DB) - as long as this is only needed within the scope of the task, and is not intended to signal to the user that the task is not currently cancellable, this will be fine.